### ISSUE TITLE
Can this be enhanced to merge into a single file in the Zip archive?

### ISSUE THREAD

**Issue by ngbalk** [gbif/hadoop-compress#1]
It looks like the current implementation will do the following:

file1 -> file1 (compressed)    \\
file2 -> file2 (compressed)  ---  files.zip (contains file1, file2, file3)
file3 -> file3 (compressed)    /

Can it be enhanced to merge these files to a single file?

file1 -> file1 (compressed)     \\
file2 -> file2 (compressed)  --- file.zip (contains a single file which is the merge of file1, file2, and file3)
file3 -> file3 (compressed)    /

**Comment by MattBlissett** [comment]
Hi,

See https://github.com/gbif/hadoop-compress/blob/master/src/test/java/org/gbif/hadoop/compress/d2/zip/ModalZipTest.java#L66 for an example, or https://github.com/gbif/occurrence/blob/master/occurrence-download/src/main/java/org/gbif/occurrence/download/file/simplecsv/SimpleCsvArchiveBuilder.java#L122 for our main usage of this in production.

This library does your second case, taking multiple compressed files and merging them (concatenating them) into a single entry (file) in the Zip archive.

**Comment by timrobertson100** [comment]
Thanks @MattBlissett @ngbalk 

In your Hadoop application, you'd add this to your CP and then enable this as the compression format. The result is that you get each file stored in the deflate2 format on HDFS. 

As you build the Zip file you just add an entry for the single file, and then append each of the parts you want in `PRE_DEFLATE` mode (i.e. they are already compressed) as per the first link Matt gives.

Please let us know how this works out for you.

**Comment by ngbalk** [comment]
Thanks for pointing me towards that example, very helpful!  

I'm also curious how this could be extended / adapted to work with an Apache Spark application?

To be a bit more specific about our use-case, we have a large DataFrame distributed across many partitions, but need to upload a single zipped file to S3.  Currently, we are achieving this by repartitioning to a single partition and THEN zipping, and uploading... but I believe we can gain some performance boosts by doing the zip step in a parallel.

**Comment by timrobertson100** [comment]
I see no reason this wouldn't work with Spark out of the box just by enabling this as a compression format and avoiding the repartition to 1. (I have not tried though)

Be aware though, that you need to do the assembly of the Zip as you pull the parts from S3 which would be fast (it's just concatenating streams, skipping each footer and then appending the proper checksums). Alternatively, you could compress in parallel, but then before you upload to S3 do the Zip assembly.

What file types are you dealing with please? CSV, Avro... ?

**Comment by ngbalk** [comment]
We are dealing with CSV.

Our code currently looks something like this:

```
      dateFrame
        .repartition(1)
        .write
        .option(codec, zipCodec)
        .csv(s3OutputPath)
```

It sounds like you're saying that if we used this compression format out of the box (without repartitioning to 1) we would effectively be uploading zip "parts" to S3 in, say, `n` parts that could then be concatenated by some other process into a single zip file.  Does this library have support for that concatenation step (i.e. "skipping each footer and appending the proper checksum"?).  

Do you have any suggestions for how to do that zip assembly before uploading to S3?

Thanks again.

**Comment by timrobertson100** [comment]
So I would _hope_ something like this would work.  We use HDFS and I'm making some assumptions of S3 here.

```
dateFrame
        .write
        .option(codec, zipCodec)
        .csv(s3OutputPath)
```

I'd expect this should create N files (one per partition) in S3. Those files are unreadable by anything, as they contain a `deflate` compressed file (same compression that Zip uses) but each file has a custom footer.

You then need to concatenate those into a real Zip file which is what [our example](https://github.com/gbif/occurrence/blob/master/occurrence-download/src/main/java/org/gbif/occurrence/download/file/simplecsv/SimpleCsvArchiveBuilder.java#L122) shows.

Your code would look something like this (Java here):

```
try (
      FSDataOutputStream zipped = targetFS.create(outputPath, true);
      ModalZipOutputStream zos = new ModalZipOutputStream(new BufferedOutputStream(zipped))
    ) {

      try {

        // Get a list of files in the S3 bucket
        List<String> compressedFiles = ...; // sorry, not sure how to

        // open a IO stream on each file and feed them into the combine stream 
        D2CombineInputStream in =
          new D2CombineInputStream(compressedFiles.stream().map(fileStatus -> {
            try {
              return sourceFS.open(fileStatus.getPath());
            } catch (IOException ex) {
              throw Throwables.propagate(ex);
            }
          }).collect(Collectors.toList()));

        // now create the Zip file noting that we instruct it the files are already DEFLATED (compressed)
        ZipEntry ze = new ZipEntry("myFile.csv").toString());
        zos.putNextEntry(ze, ModalZipOutputStream.MODE.PRE_DEFLATED); // they are already compressed
        ByteStreams.copy(in, zos);
        in.close(); // required to get the sizes
        ze.setSize(in.getUncompressedLength()); // important to set the sizes and CRC
        ze.setCompressedSize(in.getCompressedLength());
        ze.setCrc(in.getCrc32());
        zos.closeEntry();
      } catch (Exception ex) {
        // do stuff
      }
    }
```

**Comment by ngbalk** [comment]
Thanks for the example - I'll give this a try and let you know how it goes!

**Comment by ngbalk** [comment]
@timrobertson100 thought I might as well ask too - could you see a way where the "merging" step you showed above could also happen in Spark?  The simplest implementation would clearly be some simple process that streams these multiple zipped S3 chunks into the D2CombineInputStream, etc. but this essentially looks like a `.collect` operation as we collect these streams into a single stream, so I'm wondering if we can hook into the existing Spark abstraction around streaming to achieve this natively.  Thoughts?

**Comment by timrobertson100** [comment]
Interesting thought @ngbalk. This was written in Hadoop v1 days and predates Spark so what follows is just me thinking aloud, but I suspect it is doable.

Currently, it writes a set of intermediate files, then merges them efficiently using a stream process to a target file. The benefit is that we compress in parallel utilizing multiple machines. 

In spark, I wonder if we could do something like:

```
myRDD
  .mapPartitions(someCompressFunction)  // returns a tuple of (compressedData, footer)
  .repartition(1)
  .write
  .zippedCSV  // custom writer: skips the footers, concatenates the compressedData, writes a final footer
```

This would avoid the intermediate file writing step. There is always going to be a shuffle across the network either to the driver (i.e. `collect()`) but perhaps we could just use a `repartition`? An alternative could use a `mapPartitions` followed by a `collect()` avoiding the `repartition` but I don't know Spark enough to say if it could stress the driver. 

OT: I notice you are AWS. Coincidentally, the guy who sat next to me when I wrote this just started in AWS Toronto (Oliver Meyn) - he may have ideas too.

**Comment by ngbalk** [comment]
Thanks a lot - definitely an interesting idea and worth trying out.  How funny, small world!  Perhaps I'll check with Oliver too!

**Comment by timrobertson100** [comment]
I don't have time to explore this at least for the next week, but I am interested if it progresses. Would be fun if EMR ended up with Zip support based on this idea.
